Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -778,6 +795,63 @@ void Device::blit_to(Gfx::Bitmap& target)
 
     Gfx::Painter painter { target };
     painter.blit({ 0, 0 }, *m_render_target, m_render_target->rect(), 1.0f, false);
+
+    if constexpr (ENABLE_STATISTICS_OVERLAY)
+        draw_statistics_overlay(target);
+}
+
+void Device::draw_statistics_overlay(Gfx::Bitmap& target)
+{
+    static Core::ElapsedTimer timer;
+    static String debug_string;
+    static int frame_counter;
+
+    frame_counter++;
+    int milliseconds = 0;
+    if (timer.is_valid())
+        milliseconds = timer.elapsed();
+    else
+        timer.start();
+
+    Gfx::Painter painter { target };
+
+    if (milliseconds > 500) {
+
+        if (g_num_pixels == 0)
+            g_num_pixels = 1;
+
+        int num_rendertarget_pixels = m_render_target->width() * m_render_target->height();
+
+        StringBuilder builder;
+        builder.append(String::formatted("Timings      : {}ms {}FPS\n", milliseconds / frame_counter, (milliseconds > 0) ? 1000 * frame_counter / milliseconds : 9999));
+        builder.append(String::formatted("Triangles    : {}\n", g_num_rasterized_triangles));
+        builder.append(String::formatted("Pixels       : {}, Shaded: {}%, Blended: {}%, Overdraw: {}%\n",
+            g_num_pixels,
+            g_num_pixels_shaded * 100 / g_num_pixels,
+            g_num_pixels_blended * 100 / g_num_pixels,
+            g_num_pixels_shaded * 100 / num_rendertarget_pixels - 100));
+        builder.append(String::formatted("Sampler calls: {}\n", g_num_sampler_calls));
+
+        debug_string = builder.to_string();
+
+        frame_counter = 0;
+        timer.start();
+    }
+
+    g_num_rasterized_triangles = 0;
+    g_num_pixels = 0;
+    g_num_pixels_shaded = 0;
+    g_num_pixels_blended = 0;
+    g_num_sampler_calls = 0;
+
+    auto& font = Gfx::FontDatabase::default_fixed_width_font();
+
+    for (int y = -1; y < 2; y++)
+        for (int x = -1; x < 2; x++)
+            if (x != 0 && y != 0)
+                painter.draw_text(target.rect().translated(x, y), debug_string, font, Gfx::TextAlignment::TopLeft, Gfx::Color::Black);
+
+    painter.draw_text(target.rect(), debug_string, font, Gfx::TextAlignment::TopLeft, Gfx::Color::White);
```

Comment:
Maybe an esthetic improvement: give the text one or two pixels padding so it doesn't sit tightly against the window frame.

Reply:
Agreed, I moved it 2 pixels to the bottom right

Reply:
Also displaying FPS and frame time with one decimal place now.